javatests/arcs/core/analysis/EntityTypeUnionTest.kt

```diff
@@ -0,0 +1,101 @@
+
+@RunWith(JUnit4::class)
+class EntityTypeUnionTest {
+    // TODO(bgogul): hash for all Schema instances.
+
+    @Test
+    fun schemaUnionMergesNames() {
+        val emptySchemaFields = SchemaFields(emptyMap(), emptyMap())
+        val thingSchema = Schema(
+            setOf(SchemaName("Thing"), SchemaName("Another")),
+            emptySchemaFields,
+            ""
+        )
+        val objectSchema = Schema(
+            setOf(SchemaName("Object"), SchemaName("Thing")),
+            emptySchemaFields,
+            ""
+        )
+        val thingObjectSchema = requireNotNull((thingSchema union objectSchema).getOrNull())
+        thingObjectSchema?.let {
+            assertThat(it.names).containsExactly(
+                SchemaName("Thing"),
+                SchemaName("Object"),
+                SchemaName("Another")
+            )
+        }
+    }
+
+    @Test
+    fun schemaUnionMergesSingletons() {
+        val textField = SchemaFields(
+            mapOf("text" to FieldType.Text),
+            emptyMap()
+        )
+        val numberField = SchemaFields(
+            mapOf("number" to FieldType.Number),
+            emptyMap()
+        )
+        val textSchema = Schema(setOf(SchemaName("Example")), textField, "")
+        val numberSchema = Schema(setOf(SchemaName("Example")), numberField, "")
+        val result = requireNotNull((textSchema union numberSchema).getOrNull())
+        result?.let {
+            assertThat(it.names).containsExactly(SchemaName("Example"))
+            assertThat(it.fields.singletons).isEqualTo(
+                mapOf("text" to FieldType.Text, "number" to FieldType.Number)
+            )
+        }
+    }
+
+    @Test
+    fun schemaUnionDetectsIncompatibleFieldTypes() {
+        val textField = SchemaFields(
+            mapOf("num_text" to FieldType.Text),
+            emptyMap()
+        )
+        val numberField = SchemaFields(
+            mapOf("num_text" to FieldType.Number),
+            emptyMap()
+        )
+        val textSchema = Schema(setOf(SchemaName("Example")), textField, "")
+        val numberSchema = Schema(setOf(SchemaName("Example")), numberField, "")
+        with(textSchema union numberSchema) {
+            assertThat(getFailureReason()).contains("Incompatible types for field 'num_text'")
+        }
+    }
+
+    @Test
+    fun entityTypeUnionComputesUnionOfSchemas() {
+        val textField = SchemaFields(
+            mapOf("text" to FieldType.Text),
+            emptyMap()
+        )
+        val numberField = SchemaFields(
+            mapOf("number" to FieldType.Number),
+            emptyMap()
+        )
+        val textSchema = Schema(setOf(SchemaName("Example")), textField, "")
+        val numberSchema = Schema(setOf(SchemaName("Example")), numberField, "")
+        val textEntity = EntityType(textSchema)
+        val numberEntity = EntityType(numberSchema)
+        val result = requireNotNull((textEntity union numberEntity).getOrNull())
+        result?.let {
```

Comment:
This `?.` shouldn't be needed with `requireNotNull` right?

Reply:
You are right. Removed `?.`